Obtaining locale data from system (for Linux for now)

jan-hudec opened the pull request:
I've been thinking about writing localization module for Rust and since you've taken the most obvious crate name `locale` I would like to join you.

I've done localization for work project in C++ where I needed to do most things myself, because no existing code was supported across all targets. So in the end I implemented formatting support (for things we needed; we don't need dates and monetary, but we have custom rounding and support for units), reader for libintl's `.mo` files and script to extract the needed bits of locale information from [CLDR](http://cldr.unicode.org/) and storing them in the `.mo` files.

And I would like to contribute something similar to Rust. This is the first batch of changes so we can discuss the approach. The overall plan is outlined below.
# Plan

The plan is to create:
1. A locale-aware formatter that will work similarly to the `std::fmt` module, but will carry around additional object of type `locale::Locale` that will describe the locale-specific bits of formatting functionality.
2. A `locale::Locale` container that will hold individual locale facets in boxes with trait types. It will have explicit storage for the standard facets (`Numeric`, `Monetary`, `Time`, `Messages` etc.) and an option to attach custom data so users can pass additional information to formatting functions for their own types.
3. A well known instance (obtained as `Arc<Locale>`, so it can be set in a thread-safe way) that would be used for formatting by default. Setting it would not affect already existing formatters/streams (to be thread-safe).
4. Traits for the standard facets. I will in due time change the current `Numeric` and `Time` to traits and the current structs will become default implementations.
5. Standard implementations of those facets.
   a. Some facets can be universal; e.g. `Numeric`, `Time` or `Monetary` can simply hold all the necessary attributes and provide formatting function to use them. I will extend them to hold and use all attributes defined by [CLDR](http://cldr.unicode.org/).
   b. Some facets will be system specific. For example `Collate` has to use `strcoll_l` and `strxfrm_l` on Linux, but [`CompareString`](https://msdn.microsoft.com/en-us/library/windows/desktop/dd317759%28v=vs.85%29.aspx) and [`LCMapString`](https://msdn.microsoft.com/en-us/library/windows/desktop/dd318700%28v=vs.85%29.aspx) on Windows.
   c. Some facets may have alternate implementations. I will implement `Messages` facet using libintl `.mo` files, but other implementations are possible. I can imagine including [`catgets`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/catgets.html) or windows resources support in some later version.
6. Factory classes, implementing the `locale::LocaleFactory` trait, for getting the relevant data out of operating system and constructing the traits. This PR already implements this (to the extent facets are defined) for Linux and I plan to implement it for Windows (using the native API, because the libc API is very limited there) and implement extensions for KDE and GNOME (KDE has settings where parts of the locale information can be overridden). However I know nothing about MacOS X and don't have access to it, so I can't do that bit.
   
   The idea of using factories is to encapsulate the system specific code, avoid duplication (once we get the `locale_t`, we can construct all facets from it) and to allow creating combined methods for obtaining the data (e.g. the KDE factory that reads some data from the KDE config file and delegates the rest to the libc implementation). It can be hidden from the user behind default constructor in `locale::Locale`.
7. Possibly implement facets using data extracted from CLDR and bundled with the application that could be used on systems where the native data is not available like Android.
# This pull request

So far I've implemented the necessary binding to GNU LibC functions (creating `locale_t`, `nl_langinfo_l` for getting the data and `iconv` for converting them to utf-8 in case the user's locale has different codeset) and the factory traits. That you will find in this pull request.

For the time being I've left the functions `Numeric::load_user_locale` and `Time::load_user_locale` in, but they won't work on my system. I don't have the files they are trying to read (and I've generated all locales provided by the `locales` package) and don't have any files in the format they expect. For which system did you write those functions? The code from them should then end up in appropriate factory specific to that system.

jan-hudec committed:
Add myself to authors and bump version.

Also fixed the homepage link.

jan-hudec committed:
Silence some warnings by newer Rust

`env` is stable and no longer feature-gated.

jan-hudec committed:
Add trait for locale factories

There is no one way of obtaining locale data and doing locale operations:

 - On Linux functions working with the locale_t struct obtained from
   newlocale() function is the right way. They are sufficiently flexible and
   the results should be consistent with most applications.

 - But KDE has it's own additional settings. The Numeric and Monetary
   categories can be even set with parameters that don't actually match any
   existing locale. Gnome may have similar settings; I did not check yet. So
   we may want to read these _in addition to_ the glibc locale functions.

 - Windows do have most of the standard C library functions, except the
   unfortunately important nl_langinfo, but they are _not_ what is normally
   used there, they don't support UTF-8 there and they use non-standard
   locale identifiers. Instead we should use the separate native api from
   winnls.h.

 - I don't have access to MacOS, but I expect there will be some
   difference there.

 - Android does not have locale data installed on the C side at all. The only
   way to implement localization there is to take the data from CLDR and pack
   them in the application or to somehow extract them from the Java side
   which has them (from CLDR).

 - The CLDR data are more detailed than what we can extract from either
   system, so I expect some applications will want to use them anyway.

My current plan is to create a Locale structure, that will contain boxed
instances of individual components of the locale: CType, Numeric, Time,
Collate, Monetary, Messages, Paper, Name, Address, Telephone, Measurement and
ability to attach user-defined extensions that will be available in
user-defined formatting functions.

To simplify construction of the Locale class, it's constructor will take
a factory that will define appropriate sources of the information. It will be
possible to create factories that only define specific locale categories and
combine them (using the CompositeLocaleFactory) with other factories to
provide the rest.

This is the first step defining the base LocaleFactory trait.

jan-hudec committed:
Bindings for GNU LibC locale module on Linux.

On Linux the libc locale are standard method of providing locale information
and thanks to the nl_langinfo_l function all the necessary data is readily
available in a thread-safe way. We also bind iconv, which is part of libc on
Linux and can thus be conveniently used to convert from the locale charset to
Rust's internal utf-8 as we can't guarantee that user's locale is utf-8.

jan-hudec committed:
Implement default locale factory on Linux

Created function for getting the default locale factory and bound the
linux::LibCLocaleFactory to be returned on Linux. It only defines the two
facets defined so far.

jan-hudec committed:
Include the full license text

For formal correctness this is needed.

ogham commented on src/lib.rs:
```diff
@@ -21,6 +21,111 @@ use std::env::var;
+/// functionality of the underlying operating system and configured for user's default locale.
+pub fn user_locale_factory() -> SystemLocaleFactory {
+    // FIXME: Error handling? Constructing locale with "" should never fail as far as I can tell.
+    SystemLocaleFactory::new("").unwrap()
```
I had to remove the call to `.unwrap()` to get it to compile on OS X.

jan-hudec replied:
Ah, thanks. Obviously I need to connect my fork to Travis and I need to compile for Windows (I've set myself a cross toolchain), so I catch inconsistency between the system-specific implementations.

ogham commented on Cargo.toml:
```diff
@@ -1,11 +1,11 @@
 
 documentation = "http://bsago.me/doc/locale/"
-homepage = "https://github.com/ogham/locale"
+homepage = "https://github.com/ogham/rust-locale"
```
I'm so clever...

ogham commented:
Holy moly. I must apologise for taking the `locale` crate name, then, as having done localisation and CLDR work before you're probably a much better authority for it than I am! I intended to provide a much better suite of localisation functions in it, out of a sense of obligation for taking the obvious crate name, but so far I've only implemented the bare minimum to get dates and numbers working in [exa](https://github.com/ogham/exa), and I don't have experience with enough OSes to work out how they implement them. So thank you very much for writing this pull request!

I have a few thoughts on the plan, but these are mostly nitpicks or questions and I agree with what you wrote.

> A locale-aware formatter that will work similarly to the `std::fmt` module

What sort of formatting functions would we have to specify here? I thought that most of the formatting provided by `std::fmt` - things like numeric precision, decimal separators, the digits of currencies, and suchlike - would all be controlled by the locale, leaving not much left (alignment?) to be controlled by the user.

> A `locale::Locale` container that will hold individual locale facets in boxes with trait types.

> Standard implementations of those facets.

What do you mean by 'standard' here? I do have some constructors (such as `Numeric::english()`) that are used as a fallback in case the system locales fail to load.

> A well known instance (obtained as `Arc<Locale>`, so it can be set in a thread-safe way) that would be used for formatting by default.

I didn't even think of doing that - good idea!

> (e.g. the KDE factory that reads some data from the KDE config file and delegates the rest to the libc implementation)

I wasn't aware that environments like KDE provided their own localisation! What sort of things does this entail? Are there any other facets that we'd have to support?

> For which system did you write those functions? The code from them should then end up in appropriate factory specific to that system.

I wrote them for OS X, and furthermore I tested the locales in exa on a Debian VM, and it seemed to work. I tried to find information about whether the paths I was searching and the contents of the files I was looking at were standardised, but couldn't find any information on that - just information on how to override them, but not where they live by default. The code for these can be moved into the appropriate factory.

It might even be possible to get OS X to work with the libc code - I haven't tried that yet, though.

jan-hudec commented:
> What sort of formatting functions would we have to specify here? I thought that most of the formatting provided by std::fmt - things like numeric precision, decimal separators, the digits of currencies, and suchlike - would all be controlled by the locale, leaving not much left (alignment?) to be controlled by the user.

I was talking about defining a trait similar to `std::fmt::Display`, but locale-aware. It would be implemented for standard types in the locale crate.

However there are cases where user will want to define it for their own types, using the functions provided by locale as building blocks. For example application we develop at work has need to display quantities. In particular speeds, distances and sizes. So in the locale class I have additional facet for selecting between metric, imperial and US customary units (imperial uses yards, US does not) and I have types for those quantities for which special formatting functions are implemented. They use the standard numeric functions for the numbers, but on top of that they decide whether which unit to show. The value is given in the base unit and the formatting function goes from `850 m` over `2.2 km` to `57 km` depending on the value and converting to miles/yards/feet if selected in the locale. For size they can even format things like `6' 5"`. I don't want to provide this in the base library, but I want to make it reasonably easy to add. When I tried ICU, it was pretty complicated to add anything like that which was one of the reasons I decided to do it myself instead.

> What do you mean by 'standard' here? I do have some constructors (such as Numeric::english()) that are used as a fallback in case the system locales fail to load.

I mean e.g. collate using `strcoll_l` (standard) compared to collate using [Unicode collation algorithm](http://www.unicode.org/reports/tr10/) implemented in Rust. Or translation using `.mo` files (which I consider standard and want to provide as default) compared to translation reading `.xliff` or Windows resources or posix `catgets` or something.

> I wasn't aware that environments like KDE provided their own localisation! What sort of things does this entail? Are there any other facets that we'd have to support?

It's just a settings dialog that allows overriding individual bits of the numeric and monetary facets (so you can set arbitrary decimal and group separators and currency symbol and such) and specifying language (message facet) separately. It is then slapped in a config file. I suppose Gnome has something similar and puts it in gconf, but I didn't look yet. Windows also allow those settings and they don't reflect them in the libc API, which is why I want to use the WinNLS API there (when I get to it; there is sure a _lot_ of work on this).

> I wrote them for OS X

Glad to hear you have OS X, because that's the one system where I can't test beyond what can be done on Travis. We develop for iOS at work, but we only have two MacOS boxes. The guy who does the porting has one and the other is build server. So I can't just go there and play with them.

> I tested the locales in exa on a Debian VM

Strange. I have Debian and I don't have any files like `/usr/share/locale/*/LC_NUMERIC`. All the data (except translations) live in a single file `/usr/lib/locale/locale-archive` that is generated by the `localedef` utility and may change with every other version of GNU LibC.

> I tried to find information about whether the paths I was searching and the contents of the files I was looking at were standardised, but couldn't find any information on that - just information on how to override them, but not where they live by default.

The files read by libc are implementation detail and unlikely to be standardized. There are also sources for the locales, that live in `/usr/share/i18n`, that are surely described somewhere, but they require some pretty serious parsing to get the information out of them.

I am also certain that the `LC_*` variables (`LC_TIME`, `LC_NUMERIC` etc.) are supposed to name a locale (e.g. `en_US.UTF-8`) and not full path.

> It might even be possible to get OS X to work with the libc code - I haven't tried that yet, though.

Problem is, that `nl_langinfo` is only specified in POSIX, not ANSI C, `nl_langinfo_l` (to avoid global settings) is only optional and most of the item IDs for it are GNU LibC extensions. So I expect there to be some significant differences.

Ok, first I'll connect Travis to my fork and clean up the inconsistency.

jan-hudec commented:
On a side-note, bytes/kilobytes/kibibytes/megabytes/mebibytes is another example of quantity format that should be possible to incorporate in the locale-aware formatting. I don't see the setting in libc, but the KDE setting dialog has option for them with options "IEC" (KiB/MiB), "JEDEC" (KB,MB, factor 1024) or "Metric" (kB/MB, factor 1000).

jan-hudec committed:
Skip tests for locales that the host does not support.

Travis boxes don't support non-utf-8 locales (not sure they support all of
the utf-8 ones either), so the tests that need them are now wrapped in
condition for the locale successfuly constructing.

jan-hudec commented:
Ok, I have rewritten the commit that introduced the signature mismatch and added one more to disable tests for which the host does not have locales as Travis does not have some.

ogham commented:
Merged! Sorry it took me so long.

ogham commented:
Sorry, I've had a busy couple of days. Tomorrow I'll go through and adjust some things, like making the `LC_*` environment variables refer to the locale rather than the path.

Also, I have a number-prefixing package over at https://github.com/ogham/rust-number-prefix, but it doesn't do JEDEC prefixes.

jan-hudec commented:
Thanks. I'll start looking at the next steps.

> Also, I have a number-prefixing package over at https://github.com/ogham/rust-number-prefix, but it doesn't do JEDEC prefixes.

What I'd really like to see is a Rust equivalent of [Boost.Units](www.boost.org/doc/html/boost_units.html), but I fear Rust is not yet up to the task, because it uses integral template parameters which Rust can't do (hopefully yet) and I am not sure how far defining type for each number (±8 would probably be enough) would go.
